### PR TITLE
Execute task hint #2: load program

### PR DESCRIPTION
**Pull request by odesenfans**
The new `ProgramLoader` struct that takes a program and loads it into memory.

## Checklist
- [ ] Linked to Github Issue
- [x] Unit tests added
- [ ] Integration tests added.
- [ ] This change requires new documentation.
  - [ ] Documentation has been added/updated.
  - [ ] CHANGELOG has been updated.

### CODE REVIEW

File: vm/src/hint_processor/builtin_hint_processor/bootloader/types.rs [hunk 1]
```diff
@@ -22,7 +28,24 @@ impl PackedOutput {
 }
 
 #[derive(Deserialize, Debug, Clone, PartialEq)]
-pub struct Task {}
+pub struct Task {
```
Comment (notlesh) [inline vm/src/hint_processor/builtin_hint_processor/bootloader/types.rs:31]: I have a different take on `Task` here: https://github.com/Moonsong-Labs/cairo-vm/pull/22/files#diff-f3ede267dc6ba12446e91386bbf0d625a48eb269a6b8114bb4484c9f2850b448L28-L29

Maybe my `RunProgramTask` can take your `Program` as its data, e.g.

```rust
enum Task {
    RunProgramTask(Program),
    CairoPieTask(_),
}
```

File: vm/src/hint_processor/builtin_hint_processor/bootloader/program_loader.rs [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+            ProgramLoaderError::Memory(e) => HintError::Memory(e),
+            _ => HintError::CustomHint(self.to_string().into_boxed_str()),
+        }
+    }
+}
+
+/// Creates an instance of `Felt252` from a builtin name.
+///
+/// Converts the builtin name to bytes then attempts to create a felt from
+/// these bytes. This function will fail if the builtin name is over 31 characters.
```
Comment (notlesh) [inline vm/src/hint_processor/builtin_hint_processor/bootloader/program_loader.rs:31]: Is there more context you can add to this, e.g. what else uses or relies on this behavior?

File: vm/src/hint_processor/builtin_hint_processor/bootloader/program_loader.rs [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+    /// Start of the program code in the VM memory.
+    pub code_address: Relocatable,
+    /// Total size of the program in memory, header included.
+    pub size: usize,
+}
+
+/// Loads a Cairo program in the VM memory.
+pub struct ProgramLoader<'vm> {
+    /// Memory accessor.
+    memory: &'vm mut Memory,
```
Comment (notlesh) [inline vm/src/hint_processor/builtin_hint_processor/bootloader/program_loader.rs:52]: Would it simplify things to store a `Vm` instead?

**Review by notlesh — APPROVED** [review]
LGTM. I noted that `Task` is different in #22, but that can be dealt with later...

File: vm/src/hint_processor/builtin_hint_processor/bootloader/types.rs (hunk 1 above)
Comment (odesenfans) [reply vm/src/hint_processor/builtin_hint_processor/bootloader/types.rs:31]: Yeah I wanted to implement it like that as well, I'll make changes after this PR to use your model + integrate `CairoPie` in the `CairoPieTask` enum.

File: vm/src/hint_processor/builtin_hint_processor/bootloader/program_loader.rs (hunk 3 above)
Comment (odesenfans) [reply vm/src/hint_processor/builtin_hint_processor/bootloader/program_loader.rs:52]: We only have access to an `&mut VirtualMachine` anyway, so we'd still have to play with references. We only need a `&mut Memory` so I think this is the right level of abstraction. 

I hesitated to pass `&memory` as a parameter to `load_program()` instead to only need an immutable reference to `ProgramLoader`, but I decided that it was alright as the VM object is doing the same thing anyway.

**Pull request merged**